### ISSUE TITLE
Maintaining 2.5 branch

### ISSUE THREAD

**Issue by ciaranmcnulty** [phpspec/phpspec#932]
What is the best practice?

At the moment I'm suggesting bug fixes target `2.5` and new features target `master`.

Then when something's merged on `2.5` I guess it's up to the maintainer to merge that branch into `master` later. Is that per-feature, or periodically?

What's best practice?

**Comment by ciaranmcnulty** [comment]
@stof your experience will be very helpful here!

**Comment by stof** [comment]
@ciaranmcnulty doing it periodically is fine, as long as you don't wait too long to propagate a bugfix to master (waiting weeks is not good), especially given that there is no 3.0 release yet, and so there should not be that many projects using it yet.
If changes in 2.5 don't happen frequently, it may be easier to just merge into master each time, to avoid forgetting it (you can of course merge multiple bug fixes in a row and then merge 2.5 into master only once at the end, if this case happens).

I would not enforce that each merge to 2.5 is followed by a branch merge (not fine when multiple bugfixes are merged in a row). But for the case of phpspec, I think that a weekly merge would be harder to maintain than an immediate merge (Symfony is in the opposite case given that merging into older branches happens much more frequently due to the project activity rate)

**Comment by mheki** [comment]
the other issue is bugfixes touching files conflicted across versions. Example: https://github.com/phpspec/phpspec/issues/928 I left both PRs open as the code change is slightly different for `2.5` and `3.0`

**Comment by stof** [comment]
Well, this does not change things. We still need to merge 2.5 into 3.0 (otherwise, it is a totally different maintenance workflow based on rebasing, as done in Doctrine, but this is much harder to follow whether the fix was applied in all concerned branches as they are different commits).

In most cases, the maintainer can resolve conflicts quite easily.
In cases where the conflicts are hard to resolve, they can ask for a separate PR for 3.0, keep the unfixed code during the conflict resolution and then merge the PR fixing the bug again in 3.0. But this is a rare case. For reference, it happens only once every few months in Symfony.

**Comment by ciaranmcnulty** [comment]
@stof thanks for the advice, I've now merged one fix to 2.5 and then merged 2.5 into master (had to resolve some things) and it worked ok.
